ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/authorization/AsyncForwardingListener.scala

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.platform.server.api.authorization
+
+import io.grpc.ServerCall
+
+/** This listener buffers all messages until `setNextListener` is called,
+  * at which point all buffered messages are sent to the given listener.
+  * From then on, all future messages are sent directly to the given listener.
+  *
+  * The target listener is usually created through `Contexts.interceptCall` or `ServerCallHandler.startCall`.
+  * */
+abstract class AsyncForwardingListener[ReqT] extends ServerCall.Listener[ReqT] {
+  protected type Listener = ServerCall.Listener[ReqT]
+  private var stash: List[Listener => Unit] = List.empty
+  private var nextListener: Option[Listener] = None
+
+  private def enqueueOrProcess(msg: Listener => Unit): Unit = this.synchronized {
```

Comment:
It's probably safe to restrict the `synchronized` block to lock on `nextListener` rather then the whole object.

Reply:
Yes, but the only place where you would mutate the state of the list is in the `setNextListener` method. If you synchronize access to both methods using the same object there's no issue.

Before commenting on your PR I re-wrote your code (for my own thinking) like this:

```scala
abstract class AsyncForwardingListener[ReqT] extends ServerCall.Listener[ReqT] {
  private[this] var stash: mutable.ListBuffer[ServerCall.Listener[ReqT] => Unit] =
    new mutable.ListBuffer
  private[this] var nextListener: Option[ServerCall.Listener[ReqT]] = None

  private def enqueueOrProcess(msg: ServerCall.Listener[ReqT] => Unit): Unit =
    nextListener.synchronized {
      if (nextListener.isDefined) {
        msg(nextListener.get)
      } else {
        stash :+ msg
      }
    }

  protected def setNextListener(listener: ServerCall.Listener[ReqT]): Unit =
    nextListener.synchronized {
      nextListener = Some(listener)
      for (msg <- stash) {
        msg(listener)
      }
    }

  // All methods that need to be forwarded
  override def onHalfClose(): Unit = enqueueOrProcess(i => i.onHalfClose())
  override def onCancel(): Unit = enqueueOrProcess(i => i.onCancel())
  override def onComplete(): Unit = enqueueOrProcess(i => i.onComplete())
  override def onReady(): Unit = enqueueOrProcess(i => i.onReady())
  override def onMessage(message: ReqT): Unit = enqueueOrProcess(i => i.onMessage(message))
}
```

Reply:
(note that I made a mistake as the `stash` can be a `val` of course)